### ISSUE TITLE
<bit>: Is the __isa_available check for lzcnt worth the cost?

### ISSUE THREAD

**Issue by lhecker** [microsoft/STL#2849]
`std::bit_ceil` emits quite a bit of assembly on x64. This seems to occur mostly due [the branch in `_Checked_x86_x64_countl_zero`](https://github.com/microsoft/STL/blob/c34f24920e463a71791c2ee3d2bed14926518965/stl/inc/bit#L211-L212) and to a lesser extent due to [the branch in `bit_ceil`](https://github.com/microsoft/STL/blob/c34f24920e463a71791c2ee3d2bed14926518965/stl/inc/bit#L49) itself.

I've written a variant which produces a more compact result: https://godbolt.org/z/q4EEz83aW
(It also removes the extra branch on ARM64 by using conditional assignments.)

I've checked the PR that introduced the code (https://github.com/microsoft/STL/pull/795) and it appears as if the cost of this if condition wasn't discussed. The if condition generally makes sense though: `bsr` is costly on AMD CPUs (up to Zen3, 5 cycles/op latency), whereas `lzcnt` is very fast on any architecture (<= 1 cycle).
But it takes up 3 slots in the CPU's branch target buffer (contemporary hardware has ~4096 slots, newly added branches incur an extra 5-20 cycle latency), generates larger binaries after inlining and unfortunately the added instructions seem to add about ~5 cycles of latency themselves, offsetting the cost of `bsr`.

This makes me wonder: Should we drop the `__isa_available` check?

**Comment by lhecker** [comment]
I've benchmarked this on a Zen3 CPU (5950X) using google/benchmark using the following simple code:
```cpp
volatile unsigned int init;

static void BM_custom(benchmark::State& state) {
    auto v = init;
    for (auto _: state) {
        benchmark::DoNotOptimize(bit_ceil(v++));
    }
}
BENCHMARK(BM_custom);
```

Here are the results:
```
-----------------------------------------------------
Benchmark           Time             CPU   Iterations
-----------------------------------------------------
BM_std           1.50 ns         1.51 ns    560000000
BM_custom        1.06 ns         1.05 ns    640000000
```

It appears as if such a change would have a positive benefit even on the Zen3 architecture, despite `bsr` being implemented in microcode resulting in a high latency of 4 cycles ([source](https://uops.info/html-lat/ZEN3/BSR_R32_R32-Measurements.html)).

**Comment by AlexGuteniev** [comment]
#2133 previous investigation

**Comment by lhecker** [comment]
Ah, duh, I only searched for existing, _open_ issues. 🤦‍♂️

The benchmark you posted there results in some "weird" numbers for me. This is fine so far:
```
std::countl_zero: 1.360940
std::countr_zero: 0.896549

// = AVX2 ❌
// Expected to be slower on AMD, due to bsf/bsr being implemented in microcode.
std::_Countl_zero_bsr: 2.15933
std::_Countr_zero_bsf: 1.71853

// = AVX2 ✅
std::_Countl_zero_lzcnt: 0.439769    <-- 3x faster!
std::_Countr_zero_tzcnt: 0.727553    <-- only 1.2x faster - why?
```
It gets weird when I mix & match the functions, for instance:
```
std::_Countl_zero_bsr: 2.15463    <-- matched 1:1 with the expected value above
std::countr_zero:      1.29592    <-- why is suddenly slower?
```
The solution here appears to be to add a `std::this_thread::sleep_for(std::chrono::milliseconds(1000));` in between the two test runs so that the latency of the former loop doesn't affect the latter. (The benchmark only measures throughput after all.)

`lzcnt`/`tzcnt` are a _lot_ faster on AMD as expected, but why does it affect `countl_zero` so much more than `countr_zero`, despite both instructions being microcode? It's not the branch in `_Countl_zero_bsr`. If I remove that one, the duration stays the same (thanks branch predictor!).

So I plugged the test into google/benchmark: https://gist.github.com/lhecker/5ea8322a8c920312b5c2f3a7bedf06e4
Since my question isn't just about raw performance, but also about binary size and how it would work if used in a larger code base I at least added `bit_ceil` tests to the above benchmark.

This results in more stable numbers (at approx. 5GHz):
```
---------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations
---------------------------------------------------------------------
std::countl_zero                0.440 ns        0.422 ns   1000000000
std::_Countl_zero_bsr            1.05 ns         1.05 ns    640000000
std::_Countl_zero_lzcnt         0.427 ns        0.391 ns   1000000000
_Countl_zero_bsr                0.845 ns        0.837 ns    896000000
std::countr_zero                0.428 ns        0.406 ns   1000000000
std::_Countr_zero_bsf           0.845 ns        0.837 ns    896000000
std::_Countr_zero_tzcnt         0.634 ns        0.625 ns   1000000000 *
std::bit_ceil                   0.843 ns        0.854 ns    896000000
bit_ceil_optimized              0.846 ns        0.837 ns    896000000
bit_ceil_optimized_checked      0.852 ns        0.837 ns    896000000
bit_ceil_optimized_bsr           1.21 ns         1.17 ns    560000000

* consistent outlier, should be better than countr_zero 🤔
```

This makes me believe that we should make the change to `_Countl_zero_bsr` (to make it branchless) in either case - if anything it can only have a positive impact. I suspect the same to be true on Intel.
But a branchy `std::countl_zero` is still about 40% faster than the branchless `_Countl_zero_bsr` variant on Zen3. I'd be curious what the difference is on Intel but I suspect it will be a _lot_ less there, since it has fast `bsf`/`bsr`.
If anyone got time / wants to, I'd be super interested what the above benchmark results in on an Intel CPU.

@AlexGuteniev What would your choice be? Given that the `__isa_available` bloats the size of the assembly by about 2x, is it worth the 0.3ns throughput improvement on AMD? 🤔

**Comment by AlexGuteniev** [comment]
On an Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
```
2022-07-07T16:39:34+03:00
Running C:\Users\Aleksandr.Gutenev\source\repos\ConsoleApplication15\x64\Release\ConsoleApplication15.exe
Run on (12 X 2208 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 9216 KiB (x1)
-------------------------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------
std::countl_zero                0.650 ns        0.656 ns   1000000000 items_per_second=1.52381G/s
std::_Countl_zero_bsr           0.836 ns        0.837 ns    896000000 items_per_second=1.19467G/s
std::_Countl_zero_lzcnt         0.682 ns        0.688 ns   1000000000 items_per_second=1.45455G/s
_Countl_zero_bsr                 1.11 ns         1.10 ns    640000000 items_per_second=910.222M/s
std::countr_zero                0.675 ns        0.663 ns    896000000 items_per_second=1.50905G/s
std::_Countr_zero_bsf            1.04 ns         1.03 ns    640000000 items_per_second=975.238M/s
std::_Countr_zero_tzcnt         0.638 ns        0.625 ns   1000000000 items_per_second=1.6G/s
std::bit_ceil                    1.78 ns         1.76 ns    407272727 items_per_second=566.64M/s
bit_ceil_optimized               1.36 ns         1.35 ns    497777778 items_per_second=740.879M/s
bit_ceil_optimized_checked       1.42 ns         1.41 ns    497777778 items_per_second=707.951M/s
bit_ceil_optimized_bsr           1.31 ns         1.32 ns    497777778 items_per_second=758.519M/s
```

**Comment by lhecker** [comment]
A colleague kindly ran the benchmark on an i7-9700k for me with similar results:
```
---------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations
---------------------------------------------------------------------
std::countl_zero                0.570 ns        0.141 ns   1000000000
std::_Countl_zero_bsr           0.706 ns        0.125 ns   1000000000
std::_Countl_zero_lzcnt         0.538 ns        0.031 ns   1000000000
_Countl_zero_bsr                0.924 ns        0.031 ns   1000000000
std::countr_zero                0.562 ns        0.031 ns   1000000000
std::_Countr_zero_bsf           0.903 ns        0.094 ns   1000000000
std::_Countr_zero_tzcnt         0.677 ns        0.188 ns   1000000000
std::bit_ceil                    1.13 ns        0.078 ns   1000000000
bit_ceil_optimized               1.20 ns        0.203 ns   1000000000
bit_ceil_optimized_checked       1.35 ns        0.188 ns   1000000000
bit_ceil_optimized_bsr           1.13 ns        0.312 ns   1000000000
```

**Comment by AlexGuteniev** [comment]
So it looks like that micro-benchmark still shows that branch doesn't cost a lot, and its cost comparable to the unexplained variations, yet `bsr`/`lzcnt` and `bsf`/`tzcnt` difference is clearly significant and worth a branch.

**Comment by lhecker** [comment]
> So it looks like that micro-benchmark still shows that branch doesn't cost a lot, and its cost comparable to the unexplained variations

It does cost a lot: In binary size. To be specific:
| Function | binary size (Bytes) |
|----------|-------------|
| std::bit_ceil | 75 |
| bit_ceil_optimized | 87 |
| bit_ceil_optimized_checked | 58 |
| bit_ceil_optimized_bsr | 31 |

This is practically the same argument for me as for the branchless `clamp` issue I opened: Should we really optimize the STL for hot paths to the detriment of everything else? `bit_ceil_optimized` in particular shows this: By removing a branch in `std::bit_ceil` we don't reduce binary size - we increase it even further, because the optimizer at `/O2` forces even more, redundant code into the two `__isa_available` branches.

So I ask: Is a 0.3-0.5 *ns* (1-2 cycles) improvement really "worth a branch"? Worth doubling the binary size of a function? Which user of the STL does such an optimization target? It would have to be someone running a few million `countl_zero` at least before it gets noticeable, no?
Speculation will make the branch essentially free, that's true. But it's only true for small-ish hot paths like in the microbenchmarks we're running. Those 2 `countl_zero` branches will force out at least 2 branches from the target buffer. If those were important, it'll cost an extra 5-20 cycle latency to get them back in. Is the `__isa_available` check still cost free then?

Basically, to summarize, my argument is identical to `clamp`: Since the difference in performance is so miniscule, we should err on the side of compactness, because that's the safer choice.

**Comment by barcharcraz** [comment]
> Plus, speculation will eventually make that branch "essentially free" because the value of __isa_available will never change.

even so, it will take a space in the btb, and it may be cleared or evicted frequently. We could at least annotate the branch to indicate it's likely the CPU has lzcnt. It's really, really, really hard to benchmark the effects of code size.

As a practical matter the additional code size could also prevent the function from being inlined at all, which would definitely be worse than the slow version had inlining happened (esp if that then goes and prevents autovectorization)

if we get _really_ creative we could eliminate the btb entry by making the check a static branch with a "JUMP_LABEL" type construct, but that could make code size worse since it may break page sharing for that page.

We could also move the bsr verison into a function, so it won't be inlined with the lzcnt code (ofc if you actually need bsr that is rather painful).

llvm-mca results for CPUs that are slow with bsr would be interesting, since that could show some estimate of the impact of the btb entry and I$ difference.

I'm pretty warm to this idea.

**Comment by AlexGuteniev** [comment]
> Basically, to summarize, my argument is identical to `clamp`: Since the difference in performance is so miniscule, we should in my personal opinion err on the side of compactness, because that's the safer choice.

My opinion is that size impact is worth the optimization. The default `/O2` implies `/Ot`, not `/Os`, and while `/Ot` makes similar sacrifices of space for even smaller gain, it is still preferred.

Also, the main reason to prefer `cmov` clamp is stable perf for unpredicted input. Here, the checked version has this property, whereas `bsr` has a branch to check for zero input, so the unconditional bsr is the one which has surprises.

> We could also move the bsr verison into a function, so it won't be inlined with the lzcnt code (ofc if you actually need bsr that is rather painful).

Might get us something, but not too much. The call instruction itself still has a significant size.

**Comment by lhecker** [comment]
> Also, the main reason to prefer `cmov` clamp is stable perf for unpredicted input. Here, the checked version has this property, whereas `bsr` has a branch to check for zero input, so the unconditional bsr is the one which has surprises.

I have wonderful news then! 🙂
Check out my godbolt link above - it features a branchless `bsr` variant (`_Countl_zero_bsr`) which is the foundation of this issue. Without it, making `std::bit_ceil` as compact (entirely branchless) and performant as I did wouldn't have been possible (resulting in [`bit_ceil_optimized_bsr`](https://godbolt.org/z/nojnYsx83)).

Here's how it works: `_BitScanReverse` doesn't actually return a proper value, but rather the `ZF` register. x86 has no branchless `return`, but it does have a lot of branchless assignments. So we only need to make a 1 line change for it to turn branchless: https://godbolt.org/z/j5xPxGoEo

By the way, I just noticed that the benchmark gist I posted had a bug - it used `bsf` instead of `bsr`. I've updated the gist and 2 benchmark results accordingly. (Some impact to `_Countl_zero_bsr` on AMD and no impact to anything else, not even on AMD. Weird.)

**Comment by lhecker** [comment]
@StephanTLavavej I hope the summary below is alright. Let me know if I should expand on anything in particular. 🙂

> Comparisons on a CPU where BSR is slow (maybe Zen 2, we're not sure?)

All AMD CPUs have slow BSR (including Zen 3). All Intel CPUs have fast BSR (at least on par with `lzcnt`).

---

> A simplified comparison of the status quo versus the recommended branchless BSR technique

`std::countl_zero`:

CPU | Is `bsr` fast? | Before (ns/op) | Before (`int`/s) | After (ns/op) | After (`int`/s)
----|----------------|----------------|------------------|---------------|----------------
AMD 5950X | ❌ | 0.440 | 2272.73 | 1.03 (+134.09%) | 970.87 (-57.28%)
Intel i7 8750H | ✅ | 0.650 | 1538.46 | 1.11 (+70.77%) | 900.90 (-41.44%)
Intel i7 9700k | ✅ | 0.591 | 1692.05 | 0.942 (+59.39%) | 1061.57 (-37.26%)

This doesn't look particularly good, but what if we actually use the result of `std::countl_zero` like most users of the STL would? It's usually used as a building block for more complex operations after all. So a IMO more representative measurement might be...

`std::bit_ceil` (`std::countl_zero` + some simple arithmetic):

CPU | Is `bsr` fast? | Before (ns/op) | Before (`int`/s) | After (ns/op) | After (`int`/s)
----|----------------|----------------|------------------|---------------|----------------
AMD 5950X | ❌ | 0.843 | 1186.24 | 1.21 (+43.53%) | 826.45 (-30.33%)
Intel i7 8750H | ✅ | 1.78 | 561.80 | 1.31 (-26.40%) | 763.36 (+35.88%)
Intel i7 9700k | ✅ | 1.15 | 869.57 | 1.12 (-2.61%) | 892.86 (+2.68%)

Most important for me however: What about binary size?

Function | Before (Bytes) | After (Bytes)
---------|----------------|--------------
`std::countl_zero` | 24 | 18 (-25.00%)
`std::bit_ceil`    | 75 | 31 (-58.67%)

---

> Possible results from llvm-mca as @barcharcraz mentioned

I'm not familiar with how to run llvm-mca with MSVC as the compiler unfortunately.
So I went ahead, compiled it with MSVC, ported the assembly over to clang and ran it for `std::bit_ceil` on Zen 2:

   /   | Throughput (lower is better) | llvm-mca output
-------|------------------------------|----------------
Before | 9.06                         | https://godbolt.org/z/KTaefxWYs
After  | 6.79                         | https://godbolt.org/z/Ycfca1T4n

I'm not sure how useful these results are however [since](https://llvm.org/docs/CommandGuide/llvm-mca.html#instruction-flow):
> Also, llvm-mca does not model branch prediction.

**Comment by sylveon** [comment]
According to these numbers, on newer Intel CPUs the branchless BSR version only is 2.68% faster, while it results in a 30.33% throughput loss on new AMD CPUs. I do not really see a major win in going branchless, because it makes Intel processors only a sliver faster, but makes AMD CPUs a third slower.

The win in code size is nice, but if codesize is that important, code should be compiled with `/O1`. Adding a check in the implementation for `/O1` and skipping the `lzcnt` detection in that case make a lot more sense than entirely nuking it, IMO.

**Comment by lhecker** [comment]
FYI The 9700k isn't newer than the 8750H. They're both the same architecture but a different stepping with the same performance characteristics (Coffee Lake H and S). The difference is caused by the desktop vs. mobile power envelope, which allows the desktop CPU to hide the technically poorer performance of the current branchy approach by boosting higher for longer, allowing the better pipelining of the current approach to shine. Specifically, adjusted for CPU time, improvement on Intel is consistent:

CPU | Is `bsr` fast? | Before (CPU time) | After (CPU time)
----|----------------|-------------------|-----------------
AMD 5950X      | ❌ | 0.854 | 1.170 (-37.00% 📉)
Intel i7 8750H | ✅ | 1.760 | 1.320 (+25.00% 📈)
Intel i7 9700k | ✅ | 0.328 | 0.203 (+38.10% 📈)

Was there ever a decision what this STL is optimized for? Real time throughput (i.e. workstations; less constrained thermal/power envelope) or CPU time throughput (laptops; more constrained thermal/power envelope)? As seen above they aren't always identical and optimizing for one might sometimes be to the detriment of the other. I picked real time throughput as that's consistent with benchmarks done on this GitHub so far, but I'd personally go with CPU time throughput in the future.
If we want to know how it performs on modern architectures, we'd still have to test that, preferably on 11th or 12th gen Intel, since those are truly different (but I suspect the results to be similar).

**Comment by RSpliet** [comment]
Came here to see if there was already a thread on optimising the BSR-based fallback, am not disappointed.
I happen to have spent some time optimising similar code reently given how I'm working on a C++17 project that doesn't have \<bit\> yet. I found that @lhecker's suggested implementation can probably be slightly improved upon by substituting the subtraction with an XOR. Since XOR has an immediate value encoding, we save ourselves a register and an instruction. I invite you to take a look at the proposed implementation "branchless_xor()" on https://godbolt.org/z/E9d8KW4sa . You can imagine a similar implementation for the 16- and 64-bit variants.
One caveat is that we only get a CMOV substitution if at least /O1 is provided. On the surface the /O0 implementation looks no worse than the previously proposed "branchless()" implementation.

I'm afraid I don't have the right tools set up to benchmark this, so apologies for not taking further steps to try and get this over the line. That said, I hope that you can use this suggestion in your advantage!

**Comment by toughengineer** [comment]
# tl;dr:
If you want to target _any_ hardware, `__isa_available` check is justified in my opinion.\
Otherwise if you use appropriate compiler flags you just get `lzcnt`.
<hr>

It so happened that I tinkered with it a bit in the context of experimenting with integer logarithm base 2. Here's how the implementation looks taking into account @RSpliet's suggestion to use XOR instead of subtraction:
```c++
uint32_t log2_2(uint64_t v) {
  auto n = std::countl_zero(v);
  if (v == 0)
    n = 63;
  return static_cast<uint32_t>(63 ^ n);
}
```

# The `__isa_available` check
<details>
<summary>MSVC generates the following instructions: (click/tap to expand)</summary>

```asm
v$ = 8
unsigned int log2_2(unsigned __int64) PROC                             ; log2_2, COMDAT
        cmp     DWORD PTR __isa_available, 5
        mov     r8d, 63                             ; 0000003fH
        jl      SHORT $LN9@log2_2
        lzcnt   rax, rcx
        test    rcx, rcx
        cmove   eax, r8d
        xor     eax, r8d
        ret     0
$LN9@log2_2:
        bsr     rdx, rcx
        jne     SHORT $LN17@log2_2
        test    rcx, rcx
        mov     eax, 64                             ; 00000040H
        cmove   eax, r8d
        xor     eax, r8d
        ret     0
$LN17@log2_2:
        mov     eax, r8d
        sub     eax, edx
        test    rcx, rcx
        cmove   eax, r8d
        xor     eax, r8d
        ret     0
unsigned int log2_2(unsigned __int64) ENDP                             ; log2_2
```
</details>

https://godbolt.org/z/o5efP4x9n

The gist is: if `__isa_available` indicates the presence of AVX2 capability or better, we execute `lzcnt` branch, otherwise the `bsr` branch.

Counter-intuitively, the main thing that slows down this implementation is reading `__isa_available`. Instruction
```asm
cmp     DWORD PTR __isa_available, 5
```
has relatively high latency, which significantly slows down the whole function.
The evidence of that is the following: if we measure it in a loop, where the compiler puts the reading operation out of the loop during optimization (but curiously not the compare operation), it runs virtually as fast as it can (the branch is correctly predicted practically all of the time and is essentially free):
```asm
mov     r10d, DWORD PTR __isa_available ; __isa_available never changes...
...
loopstart:
...
cmp     r10d, 5          ; ...result of comparison never changes as well...
jl      SHORT $LN14@loop ; ...and this branch is essentially free in this loop
lzcnt   rax, r8
...
```

So we can think about the performance of this implementation in three types of scenarios:
* when we invoke `std::countl_zero()` _once_, in this case the performance does not really matter;
* if we use it in a loop, the compiler can move the reading of `__isa_available` out of the loop, the branch depending on it is predicted correctly virtually all of the time and is hence essentially free, we get either `lzcnt` or `bsr` version depending on the hardware we run on;
* in the worst case we execute `std::countl_zero()` without optimizations for some reason (e.g. inside a "noinline" function) and we get nominal "slow" performance (I can't say how realistic this scenario is).

# Let's compare it to what Clang and GCC generate
Clang:
```asm
log2_2(unsigned long):
        test    rdi, rdi
        je      .LBB0_1
        bsr     rcx, rdi
        xor     rcx, 63
        jmp     .LBB0_3
.LBB0_1:
        mov     ecx, 64
.LBB0_3:
        xor     ecx, 63
        xor     eax, eax
        test    rdi, rdi
        cmovne  eax, ecx
        ret
```
GCC:
```asm
log2_2(unsigned long):
        bsr     rdx, rdi
        mov     eax, 0
        cmovne  eax, edx
        ret
```
https://godbolt.org/z/594b4711M

Either way it's a variation of implementation using `bsr`. Apparently Clang and GCC developers made a trade-off of not having CPUID check in any form and just always using `bsr` when targeting _any_ hardware.

On my hardware (i5-13600k) the one with `__isa_available` check in question and the `bsr` one have _comparable_ performance. On different hardware either one may be faster but I think it is still _comparable_.

Which leads me to

# Conclusion
Considering that `lzcnt` is faster than `bsr` on older hardware where it is available (e.g. AMD Zen 3), and the fact that the compiler can optimize and make the `__isa_available` check cheap (e.g. moving it out of the loop), I think that the check is justified when you want to target _any_ hardware.

When you want ultimate performance for `std::countl_zero()` in particular, and you know what hardware it will be executed on, you can always add compiler flags:
* `-arch:AVX2` for MSVC (and Clang-cl, to target Intel Haswell and AMD Excavator CPUs and later),
* `-mabm` for GCC (to target AMD K10 and Intel Haswell CPUs and later, see below about CPUID ABM flag),
* `-march=haswell` or `-march=amdfam10` or some other capable CPU type for Clang.

In each case all the compilers just generate `lzcnt` instruction.

Or maybe use [function multiversioning](https://gcc.gnu.org/onlinedocs/gcc/Function-Multiversioning.html) (I did not investigate it).

# Other considerations
Strictly speaking to use `lzcnt` one needs to check [ABM bit via CPUID](https://en.wikipedia.org/wiki/CPUID#EAX=8000'0001h:_Extended_Processor_Info_and_Feature_Bits). By checking for AVX2 we exclude older AMD processors that had `lzcnt` capability starting with K10 (from 2007) up until Excavator (from 2015). This is less than perfect, but I wouldn't change anything to add support for almost two decade old hardware for questionable benefit.

I wouldn't use `_BitScanReverse64()` directly either because MSVC with flag `-arch:AVX2` still honestly generates `bsr`, so such code does not benefit from availability of better instructions, `lzcnt` in this case, i.e. I would use `std::countl_zero()` for cross-platform-ness and overall better result. (With `-arch:AVX2` Clang-cl generates `lzcnt` though.)

**Comment by sylveon** [comment]
FWIW, we also tried to use the fact that TZCNT and BSF encode as the same before (#1540), but because the compiler uses the zero flag, which has different behavior between both instructions, the change had to be dropped.
